File: Forge/src/main/java/hantonik/atomic/core/net/packet/IAtomicPacket.java
package hantonik.atomic.core.net.packet;

import net.minecraft.network.FriendlyByteBuf;
import net.minecraftforge.event.network.CustomPayloadEvent;

public interface IAtomicPacket<T extends IAtomicPacket<T>> {
    T decode(FriendlyByteBuf buffer);

    void encode(T packet, FriendlyByteBuf buffer);

    void onPacket(T packet, CustomPayloadEvent.Context context);
}

File: Forge/src/main/kotlin/hantonik/atomic/core/net/packet/IAtomicPacket.kt
package hantonik.atomic.core.net.packet

import net.minecraft.network.FriendlyByteBuf
import net.minecraftforge.network.NetworkEvent.Context

interface IAtomicPacket<T : IAtomicPacket<T>> {
    fun decode(buffer: FriendlyByteBuf): T

    fun encode(packet: T, buffer: FriendlyByteBuf)

    fun onPacket(packet: T, context: () -> Context)
}